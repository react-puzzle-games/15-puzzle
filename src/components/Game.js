// @ts-check

import {
  Button,
  Dialog,
  DialogActions,
  DialogContent,
  DialogTitle,
  Snackbar,
} from "@material-ui/core";
import React, { Component } from "react";
import {
  GAME_IDLE,
  GAME_OVER,
  GAME_PAUSED,
  GAME_STARTED,
} from "../lib/game-status";
import { distanceBetween, getTileCoords, invert } from "../lib/utils";
import Grid from "./Grid";
import Menu from "./Menu";

class Game extends Component {
  constructor(props) {
    super(props);

    const { numbers, tileSize, gridSize, moves = 0, seconds = 0 } = props;
    const tiles = this.generateTiles(numbers, gridSize, tileSize);

    this.state = {
      tiles,
      gameState: GAME_IDLE,
      moves,
      seconds,
      dialogOpen: false,
      snackbarOpen: false,
      snackbarText: "",
    };

    document.addEventListener("keydown", this.keyDownListener);
  }

  UNSAFE_componentWillReceiveProps(nextProps) {
    const { tileSize, gridSize } = this.props;
    const newTiles = this.generateTiles(nextProps.numbers, gridSize, tileSize);

    this.setState({
      gameState: GAME_IDLE,
      tiles: newTiles,
      moves: 0,
      seconds: 0,
    });

    clearInterval(this.timerId);
  }

  // End game by pressing CTRL + ALT + F
  keyDownListener = (key) => {
    if (key.ctrlKey && key.altKey && key.code === "KeyF") {
      const { original, gridSize, tileSize } = this.props;
      const solvedTiles = this.generateTiles(original, gridSize, tileSize).map(
        (tile, index) => {
          tile.number = index + 1;
          return Object.assign({}, tile);
        }
      );

      clearInterval(this.timerId);

      this.setState({
        gameState: GAME_OVER,
        tiles: solvedTiles,
        dialogOpen: true,
      });
    }
  };

  handleDialogClose = () => {
    this.setState({
      dialogOpen: false,
    });
  };

  handleSnackbarClose = (reason) => {
    this.setState({
      snackbarOpen: false,
    });
  };

  generateTiles(numbers, gridSize, tileSize) {
    const tiles = [];

    numbers.forEach((number, index) => {
      tiles[index] = {
        ...getTileCoords(index, gridSize, tileSize),
        width: this.props.tileSize,
        height: this.props.tileSize,
        number,
      };
    });

    return tiles;
  }

  isGameOver(tiles) {
    const correctedTiles = tiles.filter((tile) => {
      return tile.tileId + 1 === tile.number;
    });

    if (correctedTiles.length === this.props.gridSize ** 2) {
      clearInterval(this.timerId);
      return true;
    } else {
      return false;
    }
  }

  addTimer() {
    this.setState((prevState) => {
      return { seconds: prevState.seconds + 1 };
    });
  }

  setTimer() {
    this.timerId = setInterval(() => {
      this.addTimer();
    }, 1000);
  }

  onPauseClick = () => {
    this.setState((prevState) => {
      let newGameState = null;
      let newSnackbarText = null;

      if (prevState.gameState === GAME_STARTED) {
        clearInterval(this.timerId);
        newGameState = GAME_PAUSED;
        newSnackbarText = "The game is currently paused.";
      } else {
        this.setTimer();
        newGameState = GAME_STARTED;
        newSnackbarText = "Game on!";
      }

      return {
        gameState: newGameState,
        snackbarOpen: true,
        snackbarText: newSnackbarText,
      };
    });
  };

  onTileClick = (tile) => {
    if (
      this.state.gameState === GAME_OVER ||
      this.state.gameState === GAME_PAUSED
    ) {
      return;
    }

    // Set Timer in case of first click
    if (this.state.moves === 0) {
      this.setTimer();
    }

    const { gridSize } = this.props;

    // Find empty tile
    const emptyTile = this.state.tiles.find((t) => t.number === gridSize ** 2);
    const emptyTileIndex = this.state.tiles.indexOf(emptyTile);

    // Find index of tile
    const tileIndex = this.state.tiles.findIndex(
      (t) => t.number === tile.number
    );

    // Is this tale neighbouring the zero tile? If so, switch them.
    const d = distanceBetween(tile, emptyTile);
    if (d.neighbours) {
      let t = Array.from(this.state.tiles).map((t) => ({ ...t }));

      invert(t, emptyTileIndex, tileIndex, [
        "top",
        "left",
        "row",
        "column",
        "tileId",
      ]);

      const checkGameOver = this.isGameOver(t);

      this.setState({
        gameState: checkGameOver ? GAME_OVER : GAME_STARTED,
        tiles: t,
        moves: this.state.moves + 1,
        dialogOpen: checkGameOver ? true : false,
      });
    }
  };

  render() {
    const { gridSize, tileSize, onResetClick, onNewClick } = this.props;

    return (
      <div
        style={{
          display: "flex",
          flexDirection: "column",
          flexGrow: 1,
        }}
      >
        <Menu
          seconds={this.state.seconds}
          moves={this.state.moves}
          onResetClick={onResetClick}
          onPauseClick={this.onPauseClick}
          onNewClick={onNewClick}
          gameState={this.state.gameState}
        />
        <Grid
          gridSize={gridSize}
          tileSize={tileSize}
          tiles={this.state.tiles}
          onTileClick={this.onTileClick}
        />
        <Dialog open={this.state.dialogOpen} onClose={this.handleDialogClose}>
          <DialogTitle>Congratulations!</DialogTitle>
          <DialogContent>
            You've solved the puzzle in {this.state.moves} moves in{" "}
            {this.state.seconds} seconds!
          </DialogContent>
          <DialogActions>
            <Button variant="text" onClick={this.handleDialogClose}>
              Close
            </Button>
          </DialogActions>
        </Dialog>
        <Snackbar
          open={this.state.snackbarOpen}
          message={this.state.snackbarText}
          onClose={this.handleSnackbarClose}
        />
      </div>
    );
  }
}

export default Game;
