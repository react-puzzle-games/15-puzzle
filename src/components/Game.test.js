// @ts-check

import React from "react";
import renderer from "react-test-renderer";
import Game from "./Game";

test("Game renders correctly", () => {
  const tree = renderer
    .create(
      <Game
        gridSize={4}
        tileSize={90}
        numbers={[]}
        onResetClick={() => {}}
        onNewClick={() => {}}
        original={[]}
      />
    )
    .toJSON();
  expect(tree).toMatchSnapshot();
});
