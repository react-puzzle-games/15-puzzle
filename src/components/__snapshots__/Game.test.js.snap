// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Game renders correctly 1`] = `
<div
  style={
    Object {
      "display": "flex",
      "flexDirection": "column",
      "flexGrow": 1,
    }
  }
>
  <header
    className="MuiPaper-root MuiAppBar-root MuiAppBar-positionStatic MuiAppBar-colorPrimary makeStyles-root-1 MuiPaper-elevation4"
  >
    <div
      className="MuiToolbar-root MuiToolbar-regular MuiToolbar-gutters"
    >
      <div
        className="MuiTypography-root makeStyles-title-2 MuiTypography-h6"
      >
        React Puzzle Games - 15 Puzzle
      </div>
      <button
        aria-label="Start a new game"
        className="MuiButtonBase-root MuiButton-root MuiButton-text"
        disabled={false}
        onBlur={[Function]}
        onClick={[Function]}
        onDragLeave={[Function]}
        onFocus={[Function]}
        onKeyDown={[Function]}
        onKeyUp={[Function]}
        onMouseDown={[Function]}
        onMouseLeave={[Function]}
        onMouseUp={[Function]}
        onTouchEnd={[Function]}
        onTouchMove={[Function]}
        onTouchStart={[Function]}
        tabIndex={0}
        type="button"
      >
        <span
          className="MuiButton-label"
        >
          <span
            className="MuiButton-startIcon MuiButton-iconSizeMedium"
          >
            <svg
              aria-hidden={true}
              className="MuiSvgIcon-root menuIcon"
              focusable="false"
              viewBox="0 0 24 24"
            >
              <path
                d="M13 3h-2v10h2V3zm4.83 2.17l-1.42 1.42C17.99 7.86 19 9.81 19 12c0 3.87-3.13 7-7 7s-7-3.13-7-7c0-2.19 1.01-4.14 2.58-5.42L6.17 5.17C4.23 6.82 3 9.26 3 12c0 4.97 4.03 9 9 9s9-4.03 9-9c0-2.74-1.23-5.18-3.17-6.83z"
              />
            </svg>
          </span>
          <span
            className="MuiTypography-root MuiTypography-button"
          >
            New game
          </span>
        </span>
      </button>
      <button
        aria-label="Pause/Continue current game."
        className="MuiButtonBase-root MuiButton-root MuiButton-text Mui-disabled Mui-disabled"
        disabled={true}
        onBlur={[Function]}
        onClick={[Function]}
        onDragLeave={[Function]}
        onFocus={[Function]}
        onKeyDown={[Function]}
        onKeyUp={[Function]}
        onMouseDown={[Function]}
        onMouseLeave={[Function]}
        onMouseUp={[Function]}
        onTouchEnd={[Function]}
        onTouchMove={[Function]}
        onTouchStart={[Function]}
        tabIndex={-1}
        type="button"
      >
        <span
          className="MuiButton-label"
        >
          <span
            className="MuiButton-startIcon MuiButton-iconSizeMedium"
          >
            <svg
              aria-hidden={true}
              className="MuiSvgIcon-root menuIcon"
              focusable="false"
              viewBox="0 0 24 24"
            >
              <path
                d="M6 19h4V5H6v14zm8-14v14h4V5h-4z"
              />
            </svg>
          </span>
          <span
            className="MuiTypography-root MuiTypography-button"
          >
            Pause
          </span>
        </span>
      </button>
      <button
        aria-label="Reset game"
        className="MuiButtonBase-root MuiButton-root MuiButton-text"
        disabled={false}
        onBlur={[Function]}
        onClick={[Function]}
        onDragLeave={[Function]}
        onFocus={[Function]}
        onKeyDown={[Function]}
        onKeyUp={[Function]}
        onMouseDown={[Function]}
        onMouseLeave={[Function]}
        onMouseUp={[Function]}
        onTouchEnd={[Function]}
        onTouchMove={[Function]}
        onTouchStart={[Function]}
        tabIndex={0}
        type="button"
      >
        <span
          className="MuiButton-label"
        >
          <span
            className="MuiButton-startIcon MuiButton-iconSizeMedium"
          >
            <svg
              aria-hidden={true}
              className="MuiSvgIcon-root"
              focusable="false"
              viewBox="0 0 24 24"
            >
              <path
                d="M12 5V1L7 6l5 5V7c3.31 0 6 2.69 6 6s-2.69 6-6 6-6-2.69-6-6H4c0 4.42 3.58 8 8 8s8-3.58 8-8-3.58-8-8-8z"
              />
            </svg>
          </span>
          Reset game
        </span>
      </button>
      <div
        className="MuiChip-root"
        onKeyDown={[Function]}
        onKeyUp={[Function]}
      >
        <div
          className="MuiAvatar-root MuiAvatar-circular MuiChip-avatar MuiAvatar-colorDefault"
        >
          <svg
            aria-hidden={true}
            className="MuiSvgIcon-root"
            focusable="false"
            viewBox="0 0 24 24"
          >
            <path
              d="M22 5.72l-4.6-3.86-1.29 1.53 4.6 3.86L22 5.72zM7.88 3.39L6.6 1.86 2 5.71l1.29 1.53 4.59-3.85zM12.5 8H11v6l4.75 2.85.75-1.23-4-2.37V8zM12 4c-4.97 0-9 4.03-9 9s4.02 9 9 9c4.97 0 9-4.03 9-9s-4.03-9-9-9zm0 16c-3.87 0-7-3.13-7-7s3.13-7 7-7 7 3.13 7 7-3.13 7-7 7z"
            />
          </svg>
        </div>
        <span
          className="MuiChip-label"
        >
          <span
            className="MuiTypography-root MuiTypography-body1"
          >
            0
            s
          </span>
        </span>
      </div>
      <div
        className="MuiChip-root"
        onKeyDown={[Function]}
        onKeyUp={[Function]}
      >
        <div
          className="MuiAvatar-root MuiAvatar-circular MuiChip-avatar MuiAvatar-colorDefault"
        >
          <svg
            aria-hidden={true}
            className="MuiSvgIcon-root"
            focusable="false"
            viewBox="0 0 24 24"
          >
            <path
              d="M9.01 14H2v2h7.01v3L13 15l-3.99-4v3zm5.98-1v-3H22V8h-7.01V5L11 9l3.99 4z"
            />
          </svg>
        </div>
        <span
          className="MuiChip-label"
        >
          <span
            className="MuiTypography-root MuiTypography-body1"
          >
            0
          </span>
        </span>
      </div>
    </div>
  </header>
  <div
    className="makeStyles-root-3 makeStyles-root-5"
  >
    <div
      className="makeStyles-tile-4 makeStyles-tile-6"
    />
  </div>
</div>
`;
